File: day19/NotDuplicate.java
import java.util.*;

public class NotDuplicate {

    public static int singleNumber(int[] nums) {
        int result = 0;
        for (int num : nums) {
            result ^= num; // XOR cancels out pairs
        }
        return result;
    }

    public static void main(String[] args) {
        Random rand = new Random();
        int length = rand.nextInt(11) + 10; // 10 to 20 unique values

        // Generate unique values using a Set
        Set<Integer> uniqueValsSet = new HashSet<>();
        while (uniqueValsSet.size() < length) {
            uniqueValsSet.add(rand.nextInt(100));
        }

        List<Integer> uniqueVals = new ArrayList<>(uniqueValsSet);
        int single = uniqueVals.remove(uniqueVals.size() - 1); // value that appears once

        List<Integer> numsList = new ArrayList<>();
        for (int val : uniqueVals) {
            numsList.add(val);
            numsList.add(val); // add each remaining value twice
        }
        numsList.add(single); // add the one that appears once
        Collections.shuffle(numsList);

        int[] nums = numsList.stream().mapToInt(Integer::intValue).toArray();

        System.out.println("Input: " + Arrays.toString(nums));
        System.out.println("Expected single number: " + single);
        System.out.println("Computed single number: " + singleNumber(nums));
    }
}

File: day19/not_duplicate.py
import random

def single_number(nums):
    result = 0
    for num in nums:
        result ^= num  # XOR cancels out duplicates
    return result

def generate_test_case():
    length = random.randint(10, 20)
    # random.sample returns unique values (like a set)
    unique_vals = random.sample(range(0, 100), length)
    single = unique_vals.pop()  # one value will appear once
    nums = unique_vals * 2      # all others appear twice
    nums.append(single)
    random.shuffle(nums)
    return nums, single

def main():
    nums, actual_single = generate_test_case()
    print("Input:", nums)
    print("Expected single number:", actual_single)
    print("Computed single number:", single_number(nums))

if __name__ == "__main__":
    main()
